feature/issue-60: Encapsulate DyanmoDB under a single shared module

nikki-t opened the pull request:
Github Issue: #60

### Description

Encapsulate DyanmoDB under a single shared module that the API and DB modules can use for access.

### Overview of work done

Created a utlity class called Connection that creates a ServiceResource for both DynamoDB and S3. Modified API and DB code to use the Connection class for access.

### Overview of verification done

Created two unit tests to test the successful retrieval of DyanoDB and S3 service resources. Verified passed new and existing unit tests.

### Overview of integration done

Testing was completed locally and on an EC2 instance for S3 operations.

API Test Run:

```python
from hydrocron.api.controllers.timeseries import lambda_handler
event = {
    "body":{
        "feature": "Reach",
        "feature_id": "71224300701",
        "start_time": "2023-06-10T00:00:00+00:00",
        "end_time": "2023-06-10T23:59:59+00:00",
        "output": "geojson",
        "fields": "feature_id,time_str,wse,width,slope,slope2"
    }
}
data = lambda_handler(event, None)
```

API Test Run Results:

```json
{
  "status": "200 OK",
  "time": 63.917,
  "hits": 1,
  "results": {
    "csv": "",
    "geojson": {
      "type": "FeatureCollection",
      "features": [
        {
          "properties": {
            "time_str": "2023-06-10T19:39:24Z",
            "wse": "321.9042",
            "width": "3631.650517",
            "slope": "-1.86594e-05",
            "slope2": "1.050322e-05"
          },
          "geometry": {},
          "type": "Feature"
        }
      ]
    }
  }
}
```

DB Test Run:

```python
from hydrocron.db.load_data import lambda_handler
event = {
    "body": {
        "table_name": "hydrocron-swot-reach-table",
        "start_date": "2023-04-26T14:00:00",
        "end_date": "2023-04-26T15:00:00",
        "obscure_data": False
    }
}
lambda_handler(event, None)
```

DB Test Run Results:

```bash
You're now authenticated with NASA Earthdata Login
Using token with expiration date: 03/09/2024
Using environment variables for EDL
```

Confirmed data was present in local database.

## PR checklist:

* [ ] Linted
* [ ] Updated unit tests
* [ ] Updated changelog
* [ ] Integration testing

_See [Pull Request Review Checklist](../CONTRIBUTING.md#reviewing) for pointers on reviewing this pull request_

frankinspace commented:
@nikki-t I merged a couple of the other PRs; could you help resolve the conflicts?

nikki-t commented:
@frankinspace - Resolved conflicts and tested DynamoDB access locally and S3 access in an EC2 instance in "us-west-2". Existing unit tests all passed.

torimcd reviewed (commented):
This is failing in SIT with the following error:

`[ERROR] ClientError: An error occurred (AccessDeniedException) when calling the GetParameter operation: User: arn:aws:sts::206226843404:assumed-role/svc-hydrocron-sit-lambda-load-data-role/svc-hydrocron-sit-load_data-lambda is not authorized to perform: ssm:GetParameter on resource: arn:aws:ssm:us-west-2:206226843404:parameter/service/hydrocron/dynamodb_endpoint_url because no identity-based policy allows the ssm:GetParameter action`

nikki-t commented:
@torimcd When I look at the IAM role referenced in the error message I do not see an attached IAM policy that allows read access to the Hydrocron parameters. However if I look at the "lambda-execution-role" it does have a policy which allows this. Is it possible that the wrong IAM role got attached to the lambda? 

I see several roles defined in the Terraform. Maybe the "HydrocronSSMRead" policy needs to be attached to the "lambda-load-data-role"?

torimcd commented:
@nikki-t there are separate lambdas for executing the API requests (which uses the lambda-execution-role) and for loading data into dynamodb (which uses the lambda-load-data-role and lambda-load-granule-role). With the dynamodb connection now in a shared module it looks like all three lambda roles need the ssm-read-policy attached so that the connection module can be imported. If you add the ssm policy
` inline_policy {
    name   = "HydrocronSSMRead"
    policy = data.aws_iam_policy_document.ssm-read-policy.json
  }`

to the load data and load granule roles it should resolve the issue

nikki-t commented:
@torimcd - Thanks for explaining that. I see how the Lambdas are organized now. I have added the "HydrocronSSMRead" policy to the "load-granule-role" as I don't think the "load_data-lambda" requires access to the parameter store. Do you mind testing in SIT again and letting me know if there are any further issues?

torimcd commented:
@nikki-t it's still failing with the same error. I think because both the load-granule and load-data lambda handlers are in the same file, the connection to ssm gets initialized when the connections module gets imported first by the load-data lambda, even though that one doesn't use the ssm parameters. We can either add the ssm read policy to the load-data lambda as well, or split the load_data module into different files for the different lamdba handlers so the load-data lambda doesn't import the connections module.

nikki-t commented:
@torimcd That’s right! Thinking about the architecture and design of hydrocron, I like the idea of splitting out the two lambdas into separate files. But I don’t think that we are at a level of complexity where this matters. Maybe we can track it and see if it becomes necessary down the road? I have updated the Terraform to include the ssm read permissions for the load data Lambda. Do you mind testing in SIT again?

@frankinspace I updated the Changelog as well.

torimcd reviewed (approved):
This is now working in SIT, confirmed data added to db

frankinspace commented on hydrocron/api/hydrocron.py:
```diff
@@ -2,15 +2,11 @@
 Hydrocron API context module
```
I think the connection stuff was the last part of this module that was really doing anything. I think we can remove this module entirely now and just instantiate the DynamoDataRepository directly in the api controllers

nikki-t replied:
@frankinspace I have removed the api.hydrocron module and now instantiate the data repository and access in the api.timeseries module.
